packager/certcache/certcache.go

```diff
@@ -445,3 +570,171 @@ func (this *CertCache) fetchOCSP(orig []byte, ocspUpdateAfter *time.Time) []byte
+
+func (this *CertCache) getCert() *x509.Certificate {
+	if !this.hasCert() {
+		return nil
+	}
+	return this.certs[0]
+}
+
+// Set current cert with mutex protection.
+func (this *CertCache) setCerts(certs []*x509.Certificate) {
+	this.certsMu.Lock()
+	defer this.certsMu.Unlock()
+	this.certs = certs
+	this.certName = util.CertName(certs[0])
+
+	err := certloader.WriteCertsToFile(this.certs, this.CertFile)
+	if err != nil {
+		log.Printf("Unable to write certs to file: %s", this.CertFile)
+	}
+}
+
+// Set new cert with mutex protection.
+func (this *CertCache) setNewCerts(certs []*x509.Certificate) {
+	this.renewedCertsMu.Lock()
+	defer this.renewedCertsMu.Unlock()
+	this.renewedCerts = certs
+
+	if this.renewedCerts == nil {
+		err := certloader.RemoveFile(this.NewCertFile)
+		if err != nil {
+			log.Printf("Unable to remove file: %s", this.NewCertFile)
+		}
+		return
+	}
+
+	err := certloader.WriteCertsToFile(this.renewedCerts, this.NewCertFile)
+	if err != nil {
+		log.Printf("Unable to write certs to file: %s", this.NewCertFile)
+	}
+}
+
+// Update the cert in the cache if necessary.
+func (this *CertCache) updateCertIfNecessary() {
+	log.Println("Updating cert if necessary")
+	if this.certFetcher == nil {
+		// Don't request new certs from CA if certFetcher is not set. This means this instance of the amppackager
+		// is not in autorenewcert mode. Just make an attempt at reading the cert saved on disk to see if
+		// another amppackager instance that is in autorenewcert mode actually updated it with a valid cert.
+		log.Println("Certfetcher is not set, skipping cert updates. Checking cert on disk if updated.")
+		this.reloadCertIfExpired()
+		return
+	}
+	d := time.Duration(0)
+	err := errors.New("")
+	if this.hasCert() {
+		d, err = util.GetDurationToExpiry(this.certs[0], time.Now())
+	}
+	if err != nil {
+		if this.renewedCerts != nil {
+			// If renewedCerts is set, copy that over to certs
+			// and set renewedCerts to nil.
+			// TODO(banaag): do the same cert setting dance on disk.
+			// Purge OCSP cache? Make shouldUpdateOCSP() return true?
+			this.setCerts(this.renewedCerts)
+			this.setNewCerts(nil)
+			return
+		}
+		// Current cert is already invalid. Try refreshing.
+		log.Println("Warning current cert is expired, attempting to renew: ", err)
+		certs, err := this.certFetcher.FetchNewCert()
+		if err != nil {
+			log.Println("Error trying to fetch new certificates from CA: ", err)
+			return
+		}
+		this.setCerts(certs)
+		return
+	}
+	if d >= time.Duration(CertRenewalInterval) {
+		// Cert is still valid, don't do anything.
+	} else if d < time.Duration(CertRenewalInterval) {
+		// Cert is still valid, but we need to start process of requesting new cert.
+		log.Println("Warning: Current cert crossed threshold for renewal, attempting to renew.")
+		certs, err := this.certFetcher.FetchNewCert()
+		if err != nil {
+			log.Println("Error trying to fetch new certificates from CA: ", err)
+			return
+		}
+		this.setNewCerts(certs)
```

Comment:
This logic seems incorrect to me; sorry if I read it wrong. AFAICT:

1. 8 days before expiry, this code runs and creates the new cert file. Nothing uses it yet.
2. 1 day later, maintainCerts calls this function again, which will lead down here to another call to FetchNewCert. Ditto daily.
3. Finally, maintainCerts calls this function and the old cert is expired, at which point the new cert is copied on top of it.

Step 2 seems like it'll make multiple redundant ACME requests for a cert.

Step 3 seems like it's happening at the wrong time (both too soon and too late). It seems like we should wait until we can get a `GOOD` OCSP response for it, and then immediately start serving it and using it to sign SXGs, so that we don't enter that 7-day period when we can't sign SXGs that last long enough.

Reply:
Let me think about this and our offline discussion about OCSP checks and send a follow-up PR when it's done. Thanks!